File: src/org/jointheleague/memory/view/MemoryGame.java
package org.jointheleague.memory.view;

import org.jointheleague.cards.Card;
import org.jointheleague.memory.controller.GameController;
import org.jointheleague.memory.model.GameModel;

import javax.swing.*;
import java.awt.*;
import java.util.Observable;
import java.util.Observer;

public class MemoryGame extends JPanel implements Runnable, Observer {

    /**
     * Constants
     */
    private static final int NUM_ROWS = 4;
    private static final int NUM_COLUMNS = 6;
    private static final int NUM_CARDS = NUM_ROWS * NUM_COLUMNS;

    private final GameController controller;
    private final GameModel model;
    private final CardButton[] buttons;

    public MemoryGame() {
        this.model = new GameModel(NUM_CARDS);
        this.controller = new GameController(model);
        this.buttons = new CardButton[NUM_CARDS];
    }


    /**
     * Starts the game.
     */
    public static void main(String[] args) {

        SwingUtilities.invokeLater(new MemoryGame());
    }

    /**
     * Initializes the MemoryGame. Generates a list of CardButtons and places
     * them on a grid inside a JFrame.
     */
    @Override
    public void run() {
        JFrame frame = new JFrame("Memory Game");
        setLayout(new GridLayout(NUM_ROWS, NUM_COLUMNS));
        Card[] cards = model.getCards();
        for (int i = 0; i < buttons.length; i++) {
            buttons[i] = new CardButton(cards[i]);
            add(buttons[i]);
            final int buttonId = i;
            buttons[i].addActionListener(e -> controller.buttonClicked(buttonId));
        }
        frame.add(this);
        frame.pack();
        frame.addWindowListener(controller);
        frame.setVisible(true);
        model.addObserver(this);
    }


    private boolean playAnotherGame() {
        int answer = JOptionPane.showConfirmDialog(this, "Do you want to play again?", "Play again?",
                JOptionPane.YES_NO_OPTION, JOptionPane.QUESTION_MESSAGE);
        return answer == JOptionPane.YES_OPTION;
    }

    @Override
    public void update(Observable o, Object arg) {
        if (o != model) return;
        SwingUtilities.invokeLater(() ->
        {
            if (model.isNewGame()) {
                Card[] cards = model.getCards();
                for (int i = 0; i < cards.length; i++) {
                    buttons[i].setCard(cards[i]);
                }
            } else {
                boolean[] faceUp = model.getFaceUp();
                for (int i = 0; i < faceUp.length; i++) {
                    buttons[i].setFaceUp(faceUp[i]);
                }
            }
            if (model.isGameOver()) {
                controller.onGameOver(playAnotherGame());
            }
        });
    }
}

File: src/org/jointheleague/memory/view/MemoryGame.kt
package org.jointheleague.memory.view

import org.jointheleague.memory.controller.GameController
import org.jointheleague.memory.model.GAME_OVER
import org.jointheleague.memory.model.GameModel
import org.jointheleague.memory.model.NEW_GAME
import java.awt.GridLayout
import java.util.*
import javax.swing.JFrame
import javax.swing.JOptionPane
import javax.swing.JPanel
import javax.swing.SwingUtilities

/*
 * Constants
 */
private const val NUM_ROWS = 2
private const val NUM_COLUMNS = 6
private const val NUM_CARDS = NUM_ROWS * NUM_COLUMNS

class MemoryGame : JPanel(), Runnable, Observer {

    private val model: GameModel = GameModel(NUM_CARDS)
    private val controller: GameController = GameController(model)
    private val buttons: Array<CardButton> = Array(NUM_CARDS) { i ->
        CardButton(model.cards[i])
    }

    /**
     * Initializes the MemoryGame. Generates a list of CardButtons and places
     * them on a grid inside a JFrame.
     */
    override fun run() {
        val frame = JFrame("Memory Game")
        layout = GridLayout(NUM_ROWS, NUM_COLUMNS)
        for ((i, button) in buttons.withIndex()) {
            add(button)
            button.addActionListener { controller.buttonClicked(i) }
        }
        frame.add(this)
        frame.pack()
        frame.addWindowListener(controller)
        frame.isVisible = true
        model.addObserver(this)
        model.playGame()
    }


    private fun playAnotherGame(): Boolean =
            JOptionPane.YES_OPTION == JOptionPane.showConfirmDialog(
                    this,
                    "Do you want to play again?",
                    "Play again?",
                    JOptionPane.YES_NO_OPTION,
                    JOptionPane.QUESTION_MESSAGE)

    override fun update(observable: Observable, arg: Any?) {
        if (observable !== model) return
//        SwingUtilities.invokeLater {
            when (arg) {

                NEW_GAME -> {
                    for ((i, c) in model.cards.withIndex()) {
                        buttons[i].card = c
                    }
                    model.playGame()
                }

                GAME_OVER -> {
                    controller.onGameOver(playAnotherGame())
                }

                else -> for ((i, v) in model.faceUp.withIndex()) {
                    buttons[i].faceUp = v
                }
            }
//        }
    }
}

fun main() {
    SwingUtilities.invokeLater(MemoryGame())
}